### PR TITLE
Gallery viewer for additional information

### PR DESCRIPTION
**Pull request by yulieth9109**

### CODE REVIEW

**Review by Robbendebiene — CHANGES_REQUESTED** [review]
One thing that is harder to solve is that the hero transition image is overlaying the arrows that are fading in. When the image gets removed the arrows suddenly appear.

This can only be solved with https://api.flutter.dev/flutter/widgets/Hero/flightShuttleBuilder.html but please do this last after everything else is done and in a separate commit.

File: lib/widgets/gallery_viewer.dart [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    _pageController.previousPage(
+      duration: const Duration(milliseconds: 300),
+      curve: Curves.easeInOut,
+    );
+  }
+
+  void goToNextImage() {
+    _pageController.nextPage(
+      duration: const Duration(milliseconds: 300),
+      curve: Curves.easeInOut,
```
Comment (Robbendebiene) [inline lib/widgets/gallery_viewer.dart:98]:
```suggestion
      curve: Curves.easeInOutCubicEmphasized,
```

File: lib/widgets/gallery_viewer.dart [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+  void handlePageChange() {
+      tag = widget.imagesKeys[_pageController.page!.round()];
+      setState(() {});
+  }
+
+  void goToPreviousImage() {
+    _pageController.previousPage(
+      duration: const Duration(milliseconds: 300),
+      curve: Curves.easeInOut,
```
Comment (Robbendebiene) [inline lib/widgets/gallery_viewer.dart:91]:
```suggestion
      curve: Curves.easeInOutCubicEmphasized,
```

File: lib/widgets/gallery_viewer.dart [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    return Stack(
+      alignment: Alignment.center,
+      children: [
+        Positioned.fill(
+          child: Hero(
+            tag: tag,
+            child: PageView.builder(
+              itemCount: widget.images.length,
+              controller: _pageController,
+              scrollBehavior: const ScrollBehavior().copyWith(overscroll: false),
```
Comment (Robbendebiene) [inline lib/widgets/gallery_viewer.dart:119]:
```suggestion

```

File: lib/widgets/gallery_viewer.dart [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+              scrollDirection: Axis.horizontal,
+              physics: const ClampingScrollPhysics(),
```
Comment (Robbendebiene) [inline lib/widgets/gallery_viewer.dart:121]:
```suggestion

```

**Comment by Robbendebiene** [comment]
What also doesn't work well is zooming and panning in the image. Most of the time the PageView gesture wins while I think the opposite would be better (InterActiveViewer wins, and only looses when panning isn't possible any more).

Just try zooming in via pinch gesture.

File: lib/widgets/gallery_viewer.dart [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Stack(
+      alignment: Alignment.center,
+      children: [
+        Positioned.fill(
+          child: Hero(
+            tag: tag,
```
[the remark this answers is not in the document]
Comment (yulieth9109) [reply lib/widgets/gallery_viewer.dart:115]: I did that and when the user interacts with the Gallery navigation through the images once the Viewer is closed each Hero that was visualized animates. For example, the user visualizes three images then when the viewer is closed three of them animate to the small image

File: lib/widgets/gallery_viewer.dart [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+              onPressed: _pageController.hasClients && _pageController.page != null ? _pageController.page! < widget.images.length - 1 ? goToNextImage : null
+              : _pageController.initialPage < widget.images.length - 1 ? goToNextImage : null,
+              color: Theme.of(context).colorScheme.primary,
+              disabledColor: Colors.transparent,
+              icon: const Icon( Icons.arrow_circle_right, size: 30.0),
+            ),
+          ),
+        ),
+      ],
+    );
```
[the remark this answers is not in the document]
Comment (yulieth9109) [reply lib/widgets/gallery_viewer.dart:167]: @Robbendebiene could you provide what you expect with this implementation? Are the arrows animated any time the user is scrolling the images, or the animation is when the arrow should disappear?

[the remark this answers is not in the document]
Comment (Robbendebiene) [reply lib/widgets/gallery_viewer.dart:167]: So the idea is basically that the arrows fade in/out at the last/first page. But the animation is directly bound to the page controller which means slightly dragging from the last page to the previous page will also slightly start to show the right arrow.
You need to do some minor math calculations so the animation values only change from 0 to 1 whenever the first/last page is reached.

File: lib/widgets/gallery_viewer.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (Robbendebiene) [reply lib/widgets/gallery_viewer.dart:115]: Good catch. You may be able to mitigate this by destroying the pages out of view.

You could try wrapping the items of the PageView with https://api.flutter.dev/flutter/widgets/KeepAlive-class.html setting it to false

Or using https://api.flutter.dev/flutter/widgets/PageView/PageView.custom.html with a custom delegate.

File: lib/widgets/gallery_viewer.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (Robbendebiene) [reply lib/widgets/gallery_viewer.dart]:
```dart
class DerivedAnimation<T extends ChangeNotifier> extends Animation<V> {
  final T _notifier;
  final V Function(T) _transformer;

  _ValueListenableDelegateAnimation({
    required T notifier,
    required V Function(T) transformer,
  }) : 
  _notifier = notifier;
  _transformer = transformer;

  @override
  void addListener(VoidCallback listener) {
    _notifier.addListener(listener);
  }

  @override
  void addStatusListener(AnimationStatusListener listener) {
    // status will never change.
  }

  @override
  void removeListener(VoidCallback listener) {
    _notifier.removeListener(listener);
  }

  @override
  void removeStatusListener(AnimationStatusListener listener) {
    // status will never change.
  }

  @override
  AnimationStatus get status => AnimationStatus.forward;

  @override
  T get value => _transformer.call(_listenable);
}



final _rightAnimation = DerivedAnimation(_pageController, (controller) {
  return controller.page.clamp(0, 1);
});

final _X = _rightAnimation.animate(Tween());


final _leftAnimation = DerivedAnimation(_pageController, (controller) {
  return controller.page.clamp(0, 1);
});

final _Y = _leftAnimation.animate(Tween());

```

**Comment by Robbendebiene** [comment]
This currently suffers from a problem where scaling and dragging interfere.

For reference:
- https://github.com/flutter/flutter/issues/137189
- https://github.com/flutter/flutter/issues/68594
- https://github.com/flutter/flutter/issues/65006

File: lib/widgets/gallery_viewer.dart [hunk 7]
```diff
@@ -221,96 +149,126 @@ class _GalleryNavigatorState extends State<GalleryNavigator> with TickerProvider
...
+  void removeStatusListener(AnimationStatusListener listener) {
+    // status will never change.
+  }
+
+  @override
+  AnimationStatus get status => AnimationStatus.forward;
+
+  @override
+  V get value => _transformer.call(_notifier);
+}
```
Comment (Robbendebiene) [inline lib/widgets/gallery_viewer.dart:274]: Please move this to the widgets directory as `derived_animation.dart`

Ideally add a short doc comment (`/// this is a doc comment`) to the class

File: lib/widgets/gallery_viewer.dart [hunk 8]
```diff
@@ -221,96 +149,126 @@ class _GalleryNavigatorState extends State<GalleryNavigator> with TickerProvider
...
+            child: PageView.custom(
+              controller: _pageController,
+              scrollDirection: Axis.horizontal,
+              physics: _pagingEnabled ? const PageScrollPhysics() : const NeverScrollableScrollPhysics(),
+              childrenDelegate: SliverChildBuilderDelegate(
+                (context, index) {
+                return  InteractiveViewer(
+                  maxScale: 3,
+                  child: FittedBox(
+                  fit: BoxFit.contain,
```
Comment (Robbendebiene) [inline lib/widgets/gallery_viewer.dart:180]: (nit) missing indentation

File: lib/widgets/gallery_viewer.dart [hunk 9]
```diff
@@ -221,96 +149,126 @@ class _GalleryNavigatorState extends State<GalleryNavigator> with TickerProvider
...
+            onPointerUp: (event) {
+              events.clear();
+              setState(() { _pagingEnabled = true; });
+            },
+            child: PageView.custom(
+              controller: _pageController,
+              scrollDirection: Axis.horizontal,
+              physics: _pagingEnabled ? const PageScrollPhysics() : const NeverScrollableScrollPhysics(),
+              childrenDelegate: SliverChildBuilderDelegate(
+                (context, index) {
```
Comment (Robbendebiene) [inline lib/widgets/gallery_viewer.dart:176]: (nit) move this line to the previous line

**Pull request merged**